dev/financial#34 : Incorrect allocation of payment on an edited multi-line item event registration

monishdeb opened the pull request:
Overview
----------------------------------------

Edit of one line item previously led to changed allocations on more than one line item. This has been fixed so only the edited line item has new financial entries associated with it.

Steps to replicate:
1. Create new Financial Type Event Fees 2. 
2. At Administer > Financial Accounts, navigate to the Financial Account this creates also called Event Fees 2 and enter 4301 as the Accounting Code.
3. Create an event with priceset with two fields, first with financial Type Event fees, second with FT of Event Fees 2. 
4. Register in event with payment of say $25 for first ticket (Event Fees) and $10 for second ticket (Event fees 2), total $35.
5. Edit Event registration. Click Change selections. Change quantity for second field from 1 to 5, line item total of $50, contribution total of $75, total paid $35. Save. Balance is $40. 
6. Click Record Payment and pay the balance.

Before
----------------------------------------
At Contributions > Accounting Batches > Open Batches, select the batch just created and export as csv, open and view the resulting transactions: [Financial_Transactions_3_20180807210702](https://lab.civicrm.org/dev/financial/uploads/0cccae0ceba1707701f9cf699f52a3e3/Financial_Transactions_3_20180807210702.csv)

After
----------------------------------------
[Financial_Transactions_3_20181119192008.csv](https://lab.civicrm.org/dev/financial/uploads/d2d17944c0224b1bc3dff5a1880b3f21/Financial_Transactions_3_20181119192008.csv)

**RESULT**: If you compare both the CSV results, each file show same number of records in which first two records represent the financial entries for respective tickets ($25 and $10), third entry is the financial entry created to indicate the outstanding balance of $40 and is payable (i.e. is_payment = FALSE). For the last two entries the amount differ whereas now after the fix $40 is being assigned to Ticket2 (as qty is changed from 1 to 5, thus amount $10 is changed to $50) and $0 to Ticket 1 unlike earlier where amount is proportionally distributed to two tickets on basis of total contribution amount, instead of financial item's amount

Technical Details
----------------------------------------
I have changed the algorithm to do allocation on basis of financial_item.amount. Ensured that proportional allocation of financial items, in case of partial payment is not changed and we already have existing UTs for this.
    In addition I have made changes in various unit test which does not follow the standard approach to change amount of a existing contribution (in order to calculate proportional amount allocation to different line-items). 

Comments
----------------------------------------
ping @JoeMurray @kcristiano @pradpnayak @eileenmcnaughton @mattwire

JoeMurray commented:
@monishdeb failing check...

kcristiano commented:
@monishdeb I've just followed the steps detailed above on an existing event registration.

I made a change to reduce the quantity.   The event transaction looks correct (they have a quantity of 1 not 2).  However the refund was pro rated against all financial types in the contribution record as oppose to applying to the financial type that was reduced.

I also get a line item on the bookkeeping report with a Debit to Accounts Receivable and no credit account.

If I add to the event registration, I still get the odd line item with a debit to A/R but no credit account.   The additional item is recorded properly with the correct financial account and Type.

Let me know what more details i can share with you so you can replicate the issue.

cc @JoeMurray

monishdeb commented:
Thanks, @kcristiano for notifying that odd behavior.  Will let you know after the fix.

monishdeb commented:
@kcristiano @JoeMurray sorry for the delay. I have fixed the issue on refund. Here's a CSV file of financial entries created after doing a refund on event fee change - [Financial_Transactions_2_20181205023626.csv](https://lab.civicrm.org/dev/financial/uploads/f96783d0d5dea8aa99420528fa2d4151/Financial_Transactions_2_20181205023626.csv)
In the CSV:
1.  The first two rows show the respective entries of ticket 1($15x1) and ticket2 ($10x2).
2.  Later on 'Change Fee selections page' ticket 2, quantity is reduced to 1 and the 3rd entry in the CSV indicate this change of pending refund -$10.
3. Then on paying the refund of $10 via 'Record Refund' page creates two proportional entries of $0 and -$10 for respective tickets. The fourth and fifth row in the CSV represent those two entries.

Please have a look.

kcristiano commented:
@monishdeb I have applied the patch and tested again.

I am still not getting the desired results.   The refund is allocating across all financial Types in the transaction

Removing an event option (decrease):

[Original Export  ](https://lab.civicrm.org/dev/financial/uploads/45c62290aaea69ecb2d39a944bd8abb1/original.csv)

[After Edit](https://lab.civicrm.org/dev/financial/uploads/043d8c46a688b536904bace5135b969a/after.csv)

Adding an event option (decrease)

[Original Export Simple](https://lab.civicrm.org/dev/financial/uploads/aa3cb4c3ac3732c282bb1b356908b92d/oriiginal-simple.csv)

[After adding option](https://lab.civicrm.org/dev/financial/uploads/81b64ef7d244c87c5b5908755aeb8a5d/after-increase.csv)

monishdeb commented:
Hi @kcristiano, thanks for your feedback. I have tried again with decreasing the event ticket by choosing the lesser amount. And I did not found the proportional allocation of the amount across financial types. Here's a screencast, which shows the database entries, esp. the financial item records where after refund only one financial record receive the refund (-$15 here) and other $0:
![before](https://user-images.githubusercontent.com/3735621/50055633-588ecb80-0177-11e9-864e-1668e02f495c.gif)

Can you please help me replicating the refund scenario,  by mentioning the steps? Please feel free to ping me (alias @monish) in mattermost.

kcristiano commented:
The difference I can see is that I am changing the options (Price set line items) by selecting/deselecting and not changing quantities.

I can provide a copy of the database(s) I am suing to test.   Ping me if you want a download link.

monishdeb commented:
Oh ok, I have tried this scenario now. In this case, I have changed the ticket to from $100 to $50. And then refunded.  I can see now three financial transactions where the 2nd item is the reversal of 1st ticket:
![before](https://user-images.githubusercontent.com/3735621/50055997-0bf9bf00-017c-11e9-8fa0-293c055909b2.gif)

monishdeb commented:
Ok got it, I'll look into this issue and will update this PR. However  I think it would be appropriate if there is a separate sub ticket in gitlab,  because the original gitlab ticket is about fixing the incorrect proportional allocation of amounts after doing partial payment. And the new issue which we came to know here is a similar issue with the financial transaction after doing refund payment. So should I create a separate sub-ticket and therefore submit a new PR which will dependent on this PR to keep things simple and it would easier for core team to review?

kcristiano commented:
@monishdeb I don't think that https://lab.civicrm.org/dev/financial/issues/34  which is incorrect allocation is linked to partial payments at all.  At least not my reading of this issue.  The problems we have seen are due to allocations of edited items, not partial payments.

monishdeb commented:
My apologies if I am drawing the current issue into respective separate tasks based on different use-cases. As in  https://lab.civicrm.org/dev/financial/issues/34  explain a use-case where, after increasing the quantity of an item, changes the original completed contribution to a partial payment.  Then after recording that partial payment leads to incorrect allocation of payments to multiple transactions. On further testing, we found another use-case where on decreasing/changing the item to a lesser amount also leads to the same issue (after paying refund). I will treat those as a common issue which is about incorrect allocations of payment after fee-change. Will let you know after I fix that last issue. Thanks!

kcristiano commented:
@monishdeb Thank you for the further testing.   The issue we had did not involve a partial payment.  I do see where this issue is speaking to an incorrect allocation after the contribution changes to a partial payment status.  I agree that these are a common issue and based on the testing we've both done, the allocation occurs regardless of the partial payment or refund pending status.

kcristiano commented:
@monishdeb have you been able to replicate the issue we have?  Please let us know if there is anything we can do to help with this.

monishdeb commented:
Sorry @kcristiano for the delay (as I was on vacation). I have updated the PR, can you please check now?

seamuslee001 commented:
@monishdeb test fails look related https://test.civicrm.org/job/CiviCRM-Core-PR/23833/testReport/junit/(root)/CRM_Event_Form_ParticipantTest/testPaymentAllocationOnMultiLineItemEvent/ and https://test.civicrm.org/job/CiviCRM-Core-PR/23833/testReport/junit/(root)/api_v3_PaymentTest/testUpdatePayment/

kcristiano commented:
@monishdeb Thank You - I will test the functionality and report back here.   I do agree that the above test failure will need to be addressed.   But I'll focus on the functionality.

kcristiano commented:
@monishdeb I just applied the patch and attempted to remove a line item.

Rather than affect only the line item removed, the amounts were prorated.

Steps:

Register for an event with multiple options and a price set
Cancel one item
edit contribution
 

![event](https://user-images.githubusercontent.com/1892544/50697919-9f9f0c80-1011-11e9-9dd6-495dc03b8e01.png)

![contrib](https://user-images.githubusercontent.com/1892544/50697923-a168d000-1011-11e9-8bb4-09ce259cb452.png)

![bookkeeping-report](https://user-images.githubusercontent.com/1892544/50697927-a3329380-1011-11e9-9166-d98b459c0a1b.png)

JoeMurray commented:
Quoting part of original description:

> RESULT: If you compare both the CSV results, each file show same number of records in which first two records represent the financial entries for respective tickets ($25 and $10), third entry is the financial entry created to indicate the outstanding balance of $40 and is payable (i.e. is_payment = FALSE). For the last two entries the amount differ whereas now after the fix $40 is being assigned to Ticket2 (as qty is changed from 1 to 5, thus amount $10 is changed to $50) and $0 to Ticket 1 unlike earlier where amount is proportionally distributed to two tickets on basis of total contribution amount, instead of financial item's amount

It would be better if we did not create a $0 entry for a line item that is not being modified.

JoeMurray commented on CRM/Price/BAO/LineItem.php:
```diff
@@ -846,6 +846,15 @@ protected function getAdjustedFinancialItemsToRecord($entityID, $entityTable, $c
         }
         $financialItemsArray[$updateFinancialItemInfoValues['entity_id']] = $updateFinancialItemInfoValues;
       }
+      // INSERT a financial item to record 0 amount when to indicate price fee is not changed
```
If this commit merely inserts a $0 financial item record on all line items that have not had their amounts changed during a transaction, I'm not sure it is worth adding unless it also resolve other things like the inappropriate proportional allocation. It also seems like a bit of a hack to solving that problem, but if it works I'll accept it.

monishdeb commented on CRM/Price/BAO/LineItem.php:
```diff
@@ -834,7 +834,8 @@ protected function getAdjustedFinancialItemsToRecord($entityID, $entityTable, $c
       $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['html_type'] == 'Text' &&
-      $updateFinancialItemInfoValues['amount'] > 0
+      $updateFinancialItemInfoValues['amount'] > 0 &&
+      ($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['line_total'] - $totalFinancialAmount) > 0
```
This is to ensure that for the unchanged text price fee no new financial item (amount = $0) will be created.

kcristiano commented:
@monishdeb Looking much better.

So far no issues, will give full report once I complete testing

kcristiano commented:
<!-- In each category, change the word "Undecided" to "Pass" or "Issue". Add explanatory comments if prompted or desired. -->

* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) Pass
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) Pass
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) tested against 5.10 RC and master.   Edited transactions via edit contribution and Changing Event selections depebnding on the type of transaction.   All changes (additions to the event, removal of line items) processed properly.  Accounting data was allocated properly
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) Pass

kcristiano commented:
@monishdeb It looks like there is now a conflict - can you resolve?   My testing was done on the 11th.

monishdeb commented:
Thanks, @kcristiano for your feedback. I have rebased the PR and will add some new unit-tests too to capture more use-cases.

monishdeb commented:
@kcristiano earlier those test failures were surfaced after a core fn was moved to another place, fixed that now. Also added two new unit tests and all passed now :)

pradpnayak commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -5804,31 +5774,65 @@ public static function getLastFinancialItemIds($contributionId) {
+      );
+
+      // In case of partial payment, calculate proportional amount for each line-item and created respective EFT records
+      if ($contributionStatus == 'Partially paid') {
```
I think this will break if it will be a full payment.

monishdeb replied:
@pradpnayak can you please provide a use-case in steps?

pradpnayak replied:
I didn't try but  i guess it may fail for usecase
1. Create pending contribution $100
2. Record $90 payment
2. Record $10 payment --- This will set contribution status to completed and might not record an additional entry in entity financial trxn table pointing to financial item for $10

monishdeb replied:
I was able to replicate it, thank you :) Will fix it

monishdeb replied:
@pradpnayak I have fixed it and updated the PR, also the corrected the modified unit tests. Please have a look.

JoeMurray replied:
@pradpnayak we'd like to have this merged. Any remaining concerns?

pradpnayak commented on tests/phpunit/CRM/Contribute/BAO/ContributionTest.php:
```diff
@@ -1179,88 +1179,90 @@ public function testgetSalesTaxFinancialAccounts() {
    * @dataProvider getThousandSeparators
    */
-  public function testCreateProportionalEntryZeroAmount($thousandSeparator) {
+  public function testCreateProportionalFinancialEntriesZeroAmount($thousandSeparator) {
```
wasn't this test written if the there is payment of $0 then items are allocated proportionally rather updating the contribution to 0?

pradpnayak commented on tests/phpunit/api/v3/PaymentTest.php:
```diff
@@ -426,7 +426,7 @@ public function testUpdatePayment() {
     );
     $eft = $this->callAPISuccess('EntityFinancialTrxn', 'get', $params);
-    $amounts = array(66.67, 33.33);
+    $amounts = array(-33.33, -16.67);
```
I am not confident why its should be in minus when a payment is updated to 100 from 50!

kcristiano commented:
We've tested this PR in all the situations we can come up with and it is working as it should.  I had a client walk through their workflows as well.

We feel that this is good to merge.

JoeMurray commented:
@seamuslee001 @eileenmcnaughton we need someone else to merge as Kevin is our JMA client on this project.

eileenmcnaughton commented:
test this please

JoeMurray commented:
@eileenmcnaughton tests passed.

kcristiano commented:
@monishdeb @JoeMurray This has now gone stale against current master.    I've been running this on our sites and it is working as it should.  Can you rebase and I'll work to see if we can get this merged for the next RC cycle?

monishdeb commented:
@kcristiano I have rebased this PR. Waiting for CT to merge it as the test build passed earlier.

kcristiano commented:
Thanks @monishdeb Can you check the current test failure?

monishdeb commented:
Jenkins test this please

monishdeb commented:
Jenkins test this please

mlutfy commented:
This PR is causing some pretty weird errors with the tests: https://chat.civicrm.org/civicrm/pl/z1pwd9rhrjrydp8ta8senmhfta

mlutfy commented on CRM/Core/BAO/FinancialTrxn.php:
```diff
@@ -726,8 +726,8 @@ public static function updateFinancialAccountsOnPaymentInstrumentChange($inputPa
         'contribution_id' => $currentContribution->id,
-      ];
-      CRM_Contribute_BAO_Contribution::assignProportionalLineItems($trxnParams, $trxn->id, $prevContribution->total_amount);
+      );
```
@monishdeb Syntax error ^

mlutfy commented on CRM/Financial/Form/PaymentEdit.php:
```diff
@@ -214,9 +214,9 @@ protected function submit($submittedValues) {
-        ];
-        $contributionTotalAmount = CRM_Core_DAO::getFieldValue('CRM_Contribute_BAO_Contribution', $this->_contributionID, 'total_amount');
-        CRM_Contribute_BAO_Contribution::assignProportionalLineItems($trxnParams, $financialTrxn['id'], $contributionTotalAmount);
+        );
```
^

eileenmcnaughton commented:
The reason I've been staying away from this PR is the addPayments function seems like 'yet another way to add payments to a contribution that is inconsistent with other methods'. I'm not reviewing any work around payments /partial payments / financials until we resolve consolidating the back office form payment add with the api payment.create (currently stalled on this PR https://github.com/civicrm/civicrm-core/pull/13689 )

monishdeb commented:
Jenkins test this please

kcristiano commented:
@eileenmcnaughton as https://github.com/civicrm/civicrm-core/pull/13689 has been merged, what is the status of this PR.

We've had this in production for about 3 months (5.10.x) and it's critical to our clients.

kcristiano commented:
@eileenmcnaughton Just updating here based on our conversation.

The issue we have with clients is that with a price set where Financial Types/Accounts are different for each line item (which seems to be common for our use cases) the following happens on editing say an event registration

- User needs to cancel one line item, but not participation in the vent
- We cancel this via the UI - change selections for the event or cancel for the line item
- Instead of having the accounting only affect the one line item (as it should) the amount refunded/cancelled) is pro rated against all line items 

Obviously this PR must now be rebased, but beyond that, what needs to be done to be merge ready?

JoeMurray commented:
We're highly interested in moving this forward.

martinh-pw commented:
Very interested on our end as well :)

eileenmcnaughton commented:
This has some content from this PR (ie from the test) https://github.com/civicrm/civicrm-core/pull/14589

kcristiano commented:
@JoeMurray The changes through https://github.com/civicrm/civicrm-core/pull/14589#issuecomment-508591538  have made this PR "unmergable"  #14589 moves us forward. I think that we may want to look at what needs to be done based on that PR in order to move this forward

eileenmcnaughton commented:
@kcristiano @JoeMurray I think the related changes have got a clear path through to completion - I've been chipping away at . it - but only as fast as I've been able to get review on the pieces along the way - but basically we get Payment.create right and then call that from the additional payment form

eileenmcnaughton commented:
Current movement on this issue is in https://github.com/civicrm/civicrm-core/pull/14763 - agreed with @kcristiano to close this one
